### ISSUE TITLE
Non reproducibility of CLI tests

### ISSUE THREAD

**Issue by ndegory** [appcelerator-archive/amp#374]
Ref #370 

CLI tests are passing ~ half of the times.
One reason may be because the test specs files contains ordered commands that are not executed in order.

Specs:

```
    grep "^[^ ]" ./cmd/amp/cli/test_samples/09-remove-stack.yml  
    stop-stack:
    remove-stack:
    list-stack:
```

Results:

```
    cli_test.go:41: test 09-remove-stack.yml
    cli_test.go:82: [amp stack ls] Command passed.
    cli_test.go:84: [amp stack ls] Running...
    cli_test.go:85: [amp stack ls] Iteration: 1
    cli_test.go:109: [amp stack ls] Error: miss matched expectation
    cli_test.go:82: [amp stack stop stack1] Command passed.
    cli_test.go:84: [amp stack stop stack1] Running...
    cli_test.go:85: [amp stack stop stack1] Iteration: 1
    cli_test.go:113: [amp stack stop stack1] Command result:
             c36915de96d73f28347551221745e525e26643d225a9c6147348c0a300565d4c

    cli_test.go:82: [amp stack rm stack1] Command passed.
    cli_test.go:84: [amp stack rm stack1] Running...
    cli_test.go:85: [amp stack rm stack1] Iteration: 1
    cli_test.go:113: [amp stack rm stack1] Command result:
             c36915de96d73f28347551221745e525e26643d225a9c6147348c0a300565d4c
```

This is because [iterating on a map with range does not guarantee the order](https://blog.golang.org/go-maps-in-action#TOC_7.).

**Comment by generalhenry** [comment]
Why not a to level array? eg:

``` yaml
- remove-service:
  cmd: amp service rm
  args:
    - pinger
  options:
  expectation: (pinger)
- list-service:
  cmd: docker service ls
  args:
  options:
  expectation: ((ID)(\s)+(NAME)(\s)+(REPLICAS)(\s)+(IMAGE)(\s)+(COMMAND)((.)|(\s))+)
```

**Comment by subfuzion** [comment]
Many integration test runners support running tests both sequentially (ordered) and in parallel. Within our framework, it would seem natural to expect that any yaml file defines a test suite boundary for being able to run in parallel with any other test suite, and within the yaml file, without needing to complicate things each test case should just default to running sequentially as this probably aligns the most with what a test author intuited.

However, the sequence is based on the previous test exiting with a non-error exit code and doesn't mean that in fact the work was finished and that the next test can proceed, even if most of our test cases seem to be working that way right now. We probably need to introduce a few test constructs for specifying how long to wait or an api call to wait on (with corresponding regex defining the expected response, such as state information).

So providing some degree of parallelism between execution of test suites (independent yaml files) is an optimization that we can and should make. Supporting sequential execution of test cases within each test suite is also something we can do.

I'm thinking that if the top test elements are defined as an array, I would expect them to be executed sequentially (as @generalhenry pointed out). If the top level is a map, then I think it's reasonable to run them in parallel, but it's also reasonable to think that they at least start in the order that they are defined (as @bquenin already pointed out, [MapSlice](https://godoc.org/gopkg.in/yaml.v2#MapSlice) would take care of that).

**Comment by ndegory** [comment]
main issue with ordering has been fixed by #398
